### DISCUSSION TITLE
Smartwatch battery improvement using Juggluco

### DISCUSSION THREAD

**Discussion by vstoianov** [j-kaltes/Juggluco#125] [Ideas]
Hi, I have used Juggluco (Wear OS version) with Ticwatch S2 for more than 1 month. I use direct connection between the smartwatch and the sensor (Libre2). I have a battery issue - the battery cannot last more than 16-18h(Air plan mode, no notifications, WiFi disabled, Theater mode enabled, no wrist gesture, etc). If I disable the Bluetooth for 1h and then re-enable it in up to 2 minutes I can get the new sensor readings, and in this way the battery can last up to 34-40 hours which is huge improvement . My question is  ... is it possible to provide an option in the settings how frequently the readings via Bluetooth to be executed (every 2 or 3 minutes instead of every minute) ... I suppose this will improve the battery life significantly.

**Comment by j-kaltes** [comment]
No, the frequency is determined by the sensor. It sends every minute the glucose data to Juggluco if it has enabled this. Juggluco could disable it after receiving a glucose value and set an alarm 15 minutes in the future to turn it on again, this could result in receiving a glucose value every 20 minute or so.
I don't know whether this would matter much in terms of battery life. I don't believe your more than doubling of battery life. I believe that there is mainly battery consumption when you look at your watch. Receiving this glucose value every minute doesn't matter that much.
I think that you need to charge your battery two times a day or buy a second watch.

**Comment by vstoianov** [comment]
Make sense.
Can you help me with this " Juggluco could disable it after receiving a glucose value and set an alarm 15 minutes in the future to turn it on again". How can I manage this? Can you give me an example?

**Comment by vstoianov** [comment]
Today I have done a lot of tests and I have figured out that if I stop Juggluco WearOS app I can get the same battery saving in the same way like if I disable the bluetooth of the smartwatch. All that points that the drain of the battery comes from the way Juggluco uses the bluetooth or some other way to transfer some data (maybe some data sync in background). I will appreciate if you can help me to stop all connections and to set all necessary options to read the sensor(Libre 2) data without any other network activity. I did some review of the code but it will take some time to figure out what could be the problem. If I do not use Juggluco WearOS(direct link between the smartwatch and the Libre 2 sensor) the battery can last for up to 2-3 days. If I start the Juggluco WearOS ... the battery can last up to 16h. It is a little bit disappointing.

**Comment by j-kaltes** [comment]
I looked at it, but TicWatch S2 is no better smartwatch than Galaxy Watch 4. You can buy a second hands one for 40 dollar from Ebay
https://www.ebay.com/itm/175701122933.
Your expectations about battery life are ridiculous. An old WearOS watch doesn't last longer then 16 hours on a battery, except when you don't wear it.
I try to make the sensor function as best as possible with the WearOS watch and consider it terrible when it doesn’t function well, for example by losing the connection with the sensor and missing values. Some watches have the strange behavior of losing the connection after every value so that they have to renegotiate a connection after every value and thereby miss the next value. Probably this is all caused by some misplaced battery optimization by the makers of the watch.
And now you claim that you want to make your watch to function that way because you believe that that would make the battery of the watch last longer.
You shouldn’t take a WearOS watch when you want a long battery life. For that you can better take a watch that you can use with GadgetBridge or Watchdrip.

**Reply by vstoianov** [reply]
Sorry, do not misunderstand me. I have just share my observations. I did not have any issues with the smart watch and the sensor when I switched off the bluetooth or the Juggluco app to conserve the battery. But it is ridiculous to drain the battery so fast when it is all the time on. And this watch is not old one, it is new (bought on 2024  - I bought it from Mobvoi, yes, I know ... it is released 2020, but the battery of this one is in perfect condition).  Please, don't be angry! I've just share my thoughts. And I think it is not normal this app to drain the battery so fast. I will try to dive in the code and provide you a solution how to improve the communication with the sensor and to improve the battery life. We will be in touch.

**Comment by vstoianov** [comment]
Hi, after some period of "diving" in the code of the project I successfully did great battery improvement. Using the original version of Juggluco 7.6.1 for WearOS I could achieve no more than 13-14h battery life of my TicWatch S2(brand new - bought on January 2024. I am active swimmer and this watch fits best to my needs) and after the improvements which I did in the code now I can get more than 36h battery life - which is more than x2.5 improvement of the battery life of the watch!  


![juggluco-screenshot](https://github.com/j-kaltes/Juggluco/assets/5394473/32f60f5c-5743-4616-93fc-be7ad07a104a)

What was the issue:
Extremely aggressive usage of the BLE module. In your build you have all the time opened BLE connection and you have aggressive BLE scanning.
 
This is what I did to achieve the battery life improvement:
1. I have reduced the BLE scanning period to 10 sec.
2. All the time I receive glucose value from the sensor I close the GATT connection and start sync exact alarm to wait for 50 sec. When the alarm expires I call in the service related to the alarm SensorBluetooth.reconnectall(). In this way I am perfectly synchronized with the next glucose data sent by the Libre2 sensor.
3. I changed(increased) a little bit the timeout of LossOfSensorAlarm.

You can see there are no gaps in the readings:
![juggluco-watch-screenshot](https://github.com/j-kaltes/Juggluco/assets/5394473/62ed8554-e3c0-4bd7-ba91-f4fd20de9002)

 
In any case I've just want to share my experience.
You did a GREAT JOB! 
I do not underestimate what you've done!

**Comment by j-kaltes** [comment]
The connection to the sensor doesn’t work as you claim it does. Juggluco doesn’t constantly try to make a connection. It does make a connection and thereafter the initiative is on the sensor. The sensor sends every minute a notification with a glucose value.
The only situation where Juggluco tries again and again to get a connection is when making a connection failed. But this doesn’t happen that often: I for example have now a bad sensor that stops working when I make a walk, but still it was half an hour ago that it made a connection with the sensor. Half an hour the sensor sends every minute a glucose value without a connection disruption. You can see this when you look at left menu->Sensor.
Changing this so that it makes the connection again and again every minute will increase battery usage.
Your 36 minutes (EDIT: hours) on a battery charge is only possible when the watch is in watch only mode. That means that the computer in the watch isn’t running.

**Reply by vstoianov** [reply]
You are correct for some of your statements and not for the others. I will explain.
1. You are correct about that when the watch app disconnects the sensor will stop transmit the data every minute - writing to the characteristic.
2. Here you are not correct: Libre 2/3 sensor is not the active side of the communication. The Juggluco app (wearOS/mobile) is the active side and tries to connect to the sensor! When the connection is established the sensor start sending glucose value every minute - writing to the characteristic.
3. My explanation was not correct ... indeed I use an alarm + service to stop the bluetooth connection for 50 sec, and the I reconnect and wait for the first glucose value. Then I close the connection and using the alarm and the service  I reconnect and wait again.
4. The algorithm described in 3. safe the battery life. I see that you've never done something to optimize the battery life but you must think about it. 
5. I increased the battery life up to 36h, not to 36 min. Have it in mind.

I am a computer engineer for more than 25 years ... yeah ... may be a little bit old ... but smart ... and adaptive.   In the past I was some kind arrogant and did not listen the others what they say ... in time I start listening. Take it like an advice.

I can post a link to download the apk file and you will see a HUGE difference in the battery life of the watch.

Regards,

**Comment by j-kaltes** [comment]
Juggluco makes the connection and thereafter it doesn't connect again until the connection is broken (
onConnectionStateChange( state==STATE_DISCONNECTED)).
The connection can last for hours and mostly for half an hour or so. In this time the sensor sends the glucose value to Juggluco. 
Juggluco is not initiating a connection with the sensor in this time.

**Reply by vstoianov** [reply]
This is from the point of your version of the app. I did a lot of changes and all the time I disconnect from the GATT profile after successful reading of the glucose value. If somebody would like to test I can share a link from Google drive to the my version of the apk. I really spent a lot of time to make this works in my way and to improve the battery life of the watch.

**Reply by zyt** [reply]
@vstoianov could you share the code modifications that you did?
